gcode/cmds.go

```diff
@@ -49,87 +72,108 @@ type RotateZ struct {
 	Angle float64
 }
 
-func (r RotateZ) ToGCode(b *bytes.Buffer) {
+func (r RotateZ) ToGCode(b *bytes.Buffer, st State, sett GcodeSettings) State {
 	b.WriteString("G0 A" + StrF(r.Angle) + "\n")
+	return st
+
 }
 func (r RotateZ) LayersCount() int {
 	return 0
 }
 
 type LayersMoving struct {
-	Layers    []Layer
-	Index     int
-	ExtParams ExtrusionParams
+	Layers []Layer
 }
 
-func (lm LayersMoving) ToGCode(b *bytes.Buffer) {
-	eOff := 0.0
-	//reset extruder encoder
-	b.WriteString("G92 E0\n")
+func (lm LayersMoving) ToGCode(b *bytes.Buffer, state State, sett GcodeSettings) State {
 	for i := 0; i < len(lm.Layers); i++ {
-		b.WriteString(";LAYER:" + strconv.Itoa(i+lm.Index) + "\n")
-		switchFanGCode(lm.Layers[i].FanOff, b)
-		eOff = pathesToGCode(lm.Layers[i].Paths, "OUTER_PATHES", lm.Layers[i].WallPrintSpeed, lm.ExtParams, eOff, b)
-		eOff = pathesToGCode(lm.Layers[i].MiddlePs, "MIDDLE_PATHES", lm.Layers[i].PrintSpeed, lm.ExtParams, eOff, b)
-		eOff = pathesToGCode(lm.Layers[i].InnerPs, "INNER_PATHES", lm.Layers[i].PrintSpeed, lm.ExtParams, eOff, b)
-		eOff = pathesToGCode(lm.Layers[i].Fill, "FILL_PATHES", lm.Layers[i].PrintSpeed, lm.ExtParams, eOff, b)
-		eOff = decreaseEOff(eOff, b)
+		b.WriteString(";LAYER:" + strconv.Itoa(state.layerIndex) + "\n")
+		state = switchFanGCode(state, sett, b)
+
+		state.isWall = true
+		state = pathesToGCode(lm.Layers[i].Paths, "OUTER_PATHES", sett, state, b)
+		state = pathesToGCode(lm.Layers[i].MiddlePs, "MIDDLE_PATHES", sett, state, b)
+		state = pathesToGCode(lm.Layers[i].InnerPs, "INNER_PATHES", sett, state, b)
+		state.isWall = false
+		state = pathesToGCode(lm.Layers[i].Fill, "FILL_PATHES", sett, state, b)
+
+		state = decreaseEOff(state, b)
+		state.layerIndex += 1
 	}
+	return state
 }
 
 func (lm LayersMoving) LayersCount() int {
 	return len(lm.Layers)
 }
 
-func decreaseEOff(eOff float64, b *bytes.Buffer) float64 {
-	if eOff > 4000 {
+func decreaseEOff(state State, b *bytes.Buffer) State {
+	if state.eOff > 4000 {
 		b.WriteString("G92 E0\n")
-		return 0.0
-	} else {
-		return eOff
+		state.eOff = 0
 	}
+	return state
 }
 
-func switchFanGCode(fanOff bool, b *bytes.Buffer) {
-	if fanOff {
-		b.WriteString("M107\n")
+func switchFanGCode(state State, sett GcodeSettings, b *bytes.Buffer) State {
+	if state.layerIndex == 0 && sett.FanOffLayer1 {
```

Comment:
What if we desire to turn on fan after one layer, e.g. layerIndex == 1

Reply:
This setting is only about turn on/off fan on the first layer, if it will be another setting - it will be another code.